pennylane/templates/state_preparations.py

```diff
@@ -0,0 +1,330 @@
+
+    return (-1) ** sum_of_ones
+
+
+def _compute_theta(alpha):
+    """Calculates the rotation angles from the alpha vector.
+
+    Args:
+        alpha (array[float]): alpha parameters
+
+    Returns:
+        (array[float]): rotation angles theta
+    """
+    k = np.log2(alpha.shape[0])
+    factor = 2 ** (-k)
+
+    theta = sparse.dok_matrix(alpha.shape, dtype=np.float64)  # type: sparse.dok_matrix
+
+    for row in range(alpha.shape[0]):
+        # Use transpose of M:
+        entry = sum([_matrix_M_entry(col, row) * a for (col, _), a in alpha.items()])
+        entry *= factor
+        if abs(entry) > 1e-6:
+            theta[row, 0] = entry
+
+    return theta
+
+
+def _uniform_rotation_dagger(gate, alpha, control_wires, target_wire):
+    """Applies a given inverse rotation to the target qubit
+    that is uniformly controlled by the control qubits.
+
+    Args:
+        gate (qml.Operation): gate to be applied, needs to have exactly
+            one parameter
+        alpha (array[float]): alpha parameters
+        control_wires (array[int]): wires that act as control
+        target_wire (int): wire that acts as target
+    """
+
+    theta = _compute_theta(alpha)  # type: sparse.dok_matrix
+
+    gray_code_rank = len(control_wires)
+
+    if gray_code_rank == 0:
+        gate(theta[0, 0], wires=[target_wire])
+        return
+
+    code = gray_code(gray_code_rank)
+    num_selections = len(code)
+
+    control_indices = [
+        int(np.log2(int(code[i], 2) ^ int(code[(i + 1) % num_selections], 2)))
+        for i in range(num_selections)
+    ]
+
+    for i, control_index in enumerate(control_indices):
+        gate(theta[i, 0], wires=[target_wire])
+        qml.CNOT(wires=[control_wires[control_index], target_wire])
+
+
+def _uniform_rotation_z_dagger(alpha, control_wires, target_wire):
+    """Applies the inverse of a Z rotation to the target qubit
+    that is uniformly controlled by the control qubits.
+
+    Args:
+        alpha (array[float]): alpha parameters
+        control_wires (array[int]): wires that act as control
+        target_wire (int): wire that acts as target
+    """
+
+    _uniform_rotation_dagger(qml.RZ, alpha, control_wires, target_wire)
+
+
+def _uniform_rotation_y_dagger(alpha, control_wires, target_wire):
+    """Applies the inverse of a Y rotation to the target qubit
+    that is uniformly controlled by the control qubits.
+
+    Args:
+        alpha (array[float]): alpha parameters
+        control_wires (array[int]): wires that act as control
+        target_wire (int): wire that acts as target
+    """
+
+    _uniform_rotation_dagger(qml.RY, alpha, control_wires, target_wire)
+
+
+def _get_alpha_z(omega, n, k):
+    r"""Computes the rotation angles alpha for the Z rotations.
+
+    Args:
+        omega (float): phase of the input
+        n (int): total number of qubits
+        k (int): current qubit
+
+    Returns:
+        a sparse vector representing :math:`\alpha^z_k`
```

Comment:
These should have a return type
```suggestion
        scipy.sparse.dok_matrix[np.float64]: a sparse vector representing :math:`\alpha^z_k`
```